### PR TITLE
115 - include file in housekeeper-bundles by default

### PR DESCRIPTION
**Pull request by islean**
### Added

- Flag to indicate that the file should not be included in the housekeeper-bundles directory

### How to prepare for test

- [ ] ssh to hasta (depending on type of change)
- [ ] install on stage:
```bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t housekeeper -b 115-include-file-when-adding```

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: housekeeper/cli/add.py [hunk 1]
```diff
@@ -77,10 +76,23 @@ def bundle_cmd(context: click.Context, bundle_name: str, json: str):
 @click.option("-t", "--tag", "tags", multiple=True, help="tag to associate the file by")
 @click.option("-b", "--bundle-name", help="name of bundle that file should be added to")
 @click.option("-j", "--json", help="json formatted input")
+@click.option(
+    "-e",
+    "--exclude",
+    is_flag=True,
+    default=False,
+    show_default=True,
+    help="flag to not include the file in housekeeper",
```
Comment (ChrOertlin) [inline housekeeper/cli/add.py:85]:
```suggestion
    help="Flag to exclude the file from housekeeper",
```

File: housekeeper/cli/add.py [hunk 2]
```diff
@@ -106,7 +118,9 @@ def file_cmd(
 
     tags = data.get("tags", tags)
 
-    new_file = store.add_file(file_path=file_path, bundle=bundle, tags=tags)
+    new_file = store.add_file(
+        file_path=file_path, bundle=bundle, tags=tags, exclude=exclude, root=context.obj["root"]
```
Comment (ChrOertlin) [inline housekeeper/cli/add.py:122]: constant for "root"

File: housekeeper/store/api/handlers/create.py [hunk 3]
```diff
@@ -107,15 +115,24 @@ def add_file(
         self,
         file_path: Path,
         bundle: Bundle,
+        root: Path,
+        exclude: bool = False,
         to_archive: bool = False,
         tags: List[str] = None,
     ) -> File:
         """Build a new file object and add it to the latest version of an existing bundle"""
```
Comment (ChrOertlin) [inline housekeeper/store/api/handlers/create.py:123]:
```suggestion
        """Build a new file object and add it to the latest version of an existing bundle."""
```

**Review by ChrOertlin — APPROVED** [review]

File: tests/cli/add/test_cli_add_file.py [hunk 4]
```diff
@@ -83,12 +84,50 @@ def test_add_non_existing_file(base_context: Context, cli_runner: CliRunner, cap
...
+        [str(second_sample_vcf), "-b", bundle_name],
+        obj=populated_context,
+    )
+
+    # THEN assert it succedes
+    assert result.exit_code == 0
+    # THEN check that the proper information is displayed
+    assert "new file added" in caplog.text
+    # THEN check that the file has been included in the version and that the relative path is given
+    assert (housekeeper_version_dir / second_sample_vcf.name).exists()
```
[the remark this answers is not in the document]
Comment (islean) [reply tests/cli/add/test_cli_add_file.py]: Followed convention in the end

File: housekeeper/cli/add.py (hunk 2 above)
Comment (islean) [reply housekeeper/cli/add.py]: Done

File: housekeeper/store/api/handlers/create.py [hunk 5]
```diff
@@ -107,15 +115,24 @@ def add_file(
...
+        file_path_to_use: str = str(file_path.absolute())
+        if not exclude:
+            file_path_to_use = str(
+                _link_and_convert_to_relative_path(
+                    file_path=file_path, root_path=root, version=version_obj
+                )
+            )
         new_file = self.new_file(
-            path=str(file_path.absolute()),
+            path=file_path_to_use,
```
Comment (Vince-janv) [inline housekeeper/store/api/handlers/create.py:135]: The add_file function is used quite a bit in cg and I'm not super comfortable changing it. Can we create an `include_file` function that does the linking part and then invokes this function?

Comment (ChrOertlin) [reply housekeeper/store/api/handlers/create.py:135]: Yes, and that is the one we need ot streamline together with how the implentation is in housekeeper

Comment (islean) [reply housekeeper/store/api/handlers/create.py:135]: So potential fix needed in cg too?

**Comment by Vince-janv** [comment]
### Test passed
```
θ64° [vincent.janvid@hasta:~] [S_base] $ touch ~/dummy_file
θ64° [vincent.janvid@hasta:~] [S_base] $ housekeeper add file -t fastq -b test_bundle ~/dummy_file
2023-05-26 13:25:08 hasta.scilifelab.se housekeeper.cli.core[210416] INFO Use root path /home/proj/stage/housekeeper-bundles
2023-05-26 13:25:09 hasta.scilifelab.se housekeeper.cli.add[210416] INFO Running add file
2023-05-26 13:25:09 hasta.scilifelab.se housekeeper.store.api.handlers.read[210416] INFO Fetching bundle with name: test_bundle
2023-05-26 13:25:09 hasta.scilifelab.se housekeeper.include[210416] INFO linked file: /home/vincent.janvid/dummy_file -> /home/proj/stage/housekeeper-bundles/test_bundle/2023-05-25/dummy_file
2023-05-26 13:25:09 hasta.scilifelab.se housekeeper.store.api.handlers.read[210416] INFO Fetching tag with name: fastq
2023-05-26 13:25:09 hasta.scilifelab.se housekeeper.cli.add[210416] INFO new file added: test_bundle/2023-05-25/dummy_file (4010842)
θ65° [vincent.janvid@hasta:~] [S_base] 6s $ touch ~/dummy_file2
[vincent.janvid@hasta:~] [S_base] $ housekeeper add file -t fastq -b test_bundle --keep-input-path ~/dummy_file2
house2023-05-26 13:25:53 hasta.scilifelab.se housekeeper.cli.core[242009] INFO Use root path /home/proj/stage/housekeeper-bundles
k2023-05-26 13:25:54 hasta.scilifelab.se housekeeper.cli.add[242009] INFO Running add file
2023-05-26 13:25:54 hasta.scilifelab.se housekeeper.store.api.handlers.read[242009] INFO Fetching bundle with name: test_bundle
2023-05-26 13:25:54 hasta.scilifelab.se housekeeper.store.api.handlers.read[242009] INFO Fetching tag with name: fastq
2023-05-26 13:25:54 hasta.scilifelab.se housekeeper.cli.add[242009] INFO new file added: /home/vincent.janvid/dummy_file2 (4010843)
θ66° [vincent.janvid@hasta:~] [S_base] 6s $ housekeeper get bundle -v test_bundle
2023-05-26 13:26:11 hasta.scilifelab.se housekeeper.cli.core[254885] INFO Use root path /home/proj/stage/housekeeper-bundles
2023-05-26 13:26:12 hasta.scilifelab.se housekeeper.store.api.handlers.read[254885] INFO Fetching all bundles
2023-05-26 13:26:12 hasta.scilifelab.se housekeeper.store.api.handlers.read[254885] INFO Fetching bundle with name: test_bundle
                📦 Bundle table 📦
┏━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ ID     ┃ Bundle name ┃ Version IDs ┃ Created    ┃
┡━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ 138275 │ test_bundle │ 141745      │ 2023-05-25 │
└────────┴─────────────┴─────────────┴────────────┘
2023-05-26 13:26:12 hasta.scilifelab.se housekeeper.store.api.handlers.read[254885] INFO Fetching version with id: 141745
2023-05-26 13:26:12 hasta.scilifelab.se housekeeper.store.api.handlers.read[254885] INFO Fetching bundle with id: 138275
                               📕 Version table 📕
┏━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━┓
┃ ID     ┃ Bundle name ┃ Nr files ┃ Included   ┃ Archived ┃ Created    ┃ Expires ┃
┡━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━┩
│ 141745 │ test_bundle │ 6        │ 2023-05-25 │          │ 2023-05-25 │         │
└────────┴─────────────┴──────────┴────────────┴──────────┴────────────┴─────────┘
2023-05-26 13:26:12 hasta.scilifelab.se housekeeper.store.api.handlers.read[254885] INFO Fetching files from version 141745
                                         📜 Files table 📜
┏━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃ ID      ┃ File name                                                                   ┃ Tags     ┃
┡━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│ 4010814 │ /home/proj/stage/housekeeper-bundles/test_bundle/2023-05-25/test_file.txt   │ test_tag │
│ 4010815 │ /home/proj/stage/housekeeper-bundles/test_bundle/2023-05-25/test_file_2.txt │ test_tag │
│ 4010816 │ /home/proj/stage/housekeeper-bundles/test_bundle/2023-05-25/test_file_3.txt │ test_tag │
│ 4010817 │ /home/hiseq.clinical/hk_test_dir/test_file_4.txt                            │ test_tag │
│ 4010842 │ /home/proj/stage/housekeeper-bundles/test_bundle/2023-05-25/dummy_file      │ fastq    │
│ 4010843 │ /home/vincent.janvid/dummy_file2                                            │ fastq    │
└─────────┴─────────────────────────────────────────────────────────────────────────────┴──────────┘
```

File: tests/conftest.py [hunk 6]
```diff
@@ -350,6 +348,14 @@ def fixture_project_dir(tmpdir_factory) -> Path:
     shutil.rmtree(str(my_tmpdir))
 
 
+@pytest.fixture(scope="function", name="housekeeper_version_dir")
+def fixture_housekeeper_version_dir(project_dir, case_id, timestamp_string) -> Path:
+    """Path to a created directory with case and version."""
+    my_tmpdir = Path(project_dir, case_id, timestamp_string)
```
Comment (henrikstranneheim) [inline tests/conftest.py:354]:
```suggestion
    hk_version_tmpdir = Path(project_dir, case_id, timestamp_string)
```

**Review by henrikstranneheim — APPROVED** [review]
Well done 👍

File: tests/conftest.py [hunk 7]
```diff
@@ -350,6 +348,14 @@ def fixture_project_dir(tmpdir_factory) -> Path:
     shutil.rmtree(str(my_tmpdir))
 
 
+@pytest.fixture(scope="function", name="housekeeper_version_dir")
```
[the remark this answers is not in the document]
Comment (islean) [reply tests/conftest.py:351]: Given that all fixtures have "scope="function"" added, it makes more sense to me to open a small pr cleaning up the fixtures in Housekeeper than having one fixture without the scope added?

**Comment by henrikstranneheim** [comment]
@islean Great work with this. Do not forget to fix the Sonarcloud code smells before merging Prs.

**Comment by islean** [comment]
@henrikstranneheim Oh right! It is merged and deployed now but I can open a mini-pr for it and fix it tomorrow. :)

**Comment by henrikstranneheim** [comment]
Sounds great 👍

**Pull request merged**